### PR TITLE
feat!: support `oras.Fetch` and `oras.FetchBytes`

### LINKED ISSUE #272: Utility method for getting content
Sub-task of #102

### PR DESCRIPTION
**Pull request by Wwwsylvia**
Add utility methods for convenient content fetching.

Resolves #272
BREAKING CHANGE: Move `content.ErrSizeExceedLimit` to `errdef.ErrSizeExceedsLimit`
Signed-off-by: Sylvia Lei <lixlei@microsoft.com>

### CODE REVIEW

File: content.go [hunk 1]
```diff
@@ -76,25 +93,116 @@ func Resolve(ctx context.Context, target ReadOnlyTarget, ref string, opts Resolv
...
-	desc, err := target.Resolve(ctx, ref)
+	desc, err := target.Resolve(ctx, reference)
 	if err != nil {
 		return ocispec.Descriptor{}, err
 	}
+	return platform.SelectManifest(ctx, target, desc, opts.TargetPlatform)
+}
+
+// Fetch fetches the content identified by the reference.
+func Fetch(ctx context.Context, target ReadOnlyTarget, reference string) (ocispec.Descriptor, io.ReadCloser, error) {
```
Comment (Wwwsylvia) [inline content.go:117]: I'm wondering if this is necessary. If so, it looks like we should also provide a platform option.

File: content.go [hunk 2]
```diff
@@ -76,25 +93,116 @@ func Resolve(ctx context.Context, target ReadOnlyTarget, ref string, opts Resolv
...
+// FetchContentOptions contains parameters for oras.FetchContent.
+type FetchContentOptions struct {
+	// ResolveOptions contains parameters for resolving reference.
+	ResolveOptions
+	// SizeLimit limits the max size of the fetched content.
+	SizeLimit int64
+}
+
+// FetchContent fetches the content bytes identified by the reference.
+func FetchContent(ctx context.Context, target ReadOnlyTarget, reference string, opts FetchContentOptions) (ocispec.Descriptor, []byte, error) {
```
Comment (Wwwsylvia) [inline content.go:147]: Any naming suggestions are welcomed!

File: content/utils.go [hunk 3]
```diff
@@ -62,3 +66,36 @@ func ReadAll(r io.Reader, desc ocispec.Descriptor) ([]byte, error) {
...
+		return nil, err
+	}
+	defer rc.Close()
+	return ReadAll(rc, desc)
+}
+
+// FetchAllWithLimit safely fetches the content described by the descriptor.
+// The fetched content is verified against the size and the digest.
+// The size of the fetched content cannot exceed the given size limit.
+func FetchAllWithLimit(ctx context.Context, fetcher Fetcher, desc ocispec.Descriptor, limit int64) ([]byte, error) {
```
Comment (Wwwsylvia) [inline content/utils.go:84]: Or `LimitedFetchAll`?

File: content.go (hunk 1 above)
Comment (Wwwsylvia) [reply content.go:117]: OK let me refactor this.

File: content.go [hunk 4]
```diff
@@ -76,25 +90,106 @@ func Resolve(ctx context.Context, target ReadOnlyTarget, ref string, opts Resolv
...
+type FetchBytesOptions struct {
+	FetchOptions
+	// SizeLimit limits the max size of the fetched content.
+	// If SizeLimit is not specified, or the specified value is less than or
+	// equal to 0, it will be considered as infinity.
+	SizeLimit int64
+}
+
+// FetchBytes fetches the content bytes identified by the reference.
+func FetchBytes(ctx context.Context, target ReadOnlyTarget, reference string, opts FetchBytesOptions) (ocispec.Descriptor, []byte, error) {
```
Comment (Wwwsylvia) [inline content.go:170]: Any naming suggestions are welcomed!

File: content.go [hunk 5]
```diff
@@ -60,14 +63,25 @@ type ResolveOptions struct {
...
+	}
+
+	proxy := cas.NewProxy(target, cas.NewMemory())
+	return resolve(ctx, target, proxy, reference, opts)
+}
+
+// resolve resolves a descriptor with provided reference from the target, with
+// specified caching.
+func resolve(ctx context.Context, target ReadOnlyTarget, proxy *cas.Proxy, reference string, opts ResolveOptions) (ocispec.Descriptor, error) {
+	if opts.TargetPlatform == nil {
```
Comment (lizMSFT) [inline content.go:78]: Do we still need to check `opts.TargetPlatform == nil` here? I noticed both function `Resolve` and function `Fetch` have already checked `TargetPlatform == nil` before calling this `resolve` function

File: content.go [hunk 6]
```diff
@@ -60,14 +63,25 @@ type ResolveOptions struct {
...
 
 // Resolve resolves a descriptor with provided reference from the target.
-func Resolve(ctx context.Context, target ReadOnlyTarget, ref string, opts ResolveOptions) (ocispec.Descriptor, error) {
+func Resolve(ctx context.Context, target ReadOnlyTarget, reference string, opts ResolveOptions) (ocispec.Descriptor, error) {
 	if opts.TargetPlatform == nil {
-		return target.Resolve(ctx, ref)
+		return target.Resolve(ctx, reference)
+	}
+
+	proxy := cas.NewProxy(target, cas.NewMemory())
```
Comment (shizhMSFT) [inline content.go:71]: Why do we need a cache proxy here for all code path?

File: content.go [hunk 7]
```diff
@@ -76,25 +90,106 @@ func Resolve(ctx context.Context, target ReadOnlyTarget, ref string, opts Resolv
...
+	var bytes []byte
+	if opts.SizeLimit > 0 {
+		if desc.Size > opts.SizeLimit {
+			return ocispec.Descriptor{}, nil, fmt.Errorf(
+				"content size %v exceeds max size limit %v: %w",
+				desc.Size,
+				opts.SizeLimit,
+				errdef.ErrSizeExceedsLimit)
+		}
+		bytes, err = content.ReadAll(io.LimitReader(rc, opts.SizeLimit), desc)
```
Comment (shizhMSFT) [inline content.go:186]: `content.ReadAll` takes care of the size limit.
```suggestion
		bytes, err = content.ReadAll(rc, desc)
```

File: content.go (hunk 5 above)
Comment (Wwwsylvia) [reply content.go:78]: Right, we don't need it. 😀

File: content.go (hunk 6 above)
Comment (Wwwsylvia) [reply content.go:71]: Nice catch!

**Review by shizhMSFT — APPROVED** [review]
LGTM

File: internal/platform/platform.go [hunk 8]
```diff
@@ -61,3 +70,67 @@ func isSubset(a, b []string) bool {
...
+			return ocispec.Descriptor{}, err
+		}
+
+		// platform filter
+		for _, m := range manifests {
+			if Match(m.Platform, p) {
+				return m, nil
+			}
+		}
+		return ocispec.Descriptor{}, errdef.ErrNotFound
```
[the remark this answers is not in the document]
Comment (Wwwsylvia) [reply internal/platform/platform.go:91]: Will fix in a separate PR. Please track #290.

**Pull request merged**